pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderTokenOIDC.java

```diff
@@ -0,0 +1,372 @@
+    }
+
+    @Override
+    public String getAuthMethodName() {
+        return TOKEN;
+    }
+
+    @Override
+    public String authenticate(AuthenticationDataSource authData) throws AuthenticationException {
+        try {
+            // Get Token
+            String token;
+            String role;
+            token = getToken(authData);
+
+            // Parse Token by validating
+
+
+            role = getPrincipal(authenticateToken(token));
+
+
+            AuthenticationMetrics.authenticateSuccess(getClass().getSimpleName(), getAuthMethodName());
+            return role;
+        } catch (AuthenticationException exception) {
+            AuthenticationMetrics.authenticateFailure(getClass().getSimpleName(), getAuthMethodName(), exception.getMessage());
+            throw exception;
+        }
+    }
+
+    @Override
+    public AuthenticationState newAuthState(AuthData authData, SocketAddress remoteAddress, SSLSession sslSession)
+            throws AuthenticationException {
+        return new TokenAuthenticationState(this, authData, remoteAddress, sslSession);
+    }
+
+    public static String getToken(AuthenticationDataSource authData) throws AuthenticationException {
+        if (authData.hasDataFromCommand()) {
+            // Authenticate Pulsar binary connection
+            return validateToken(authData.getCommandData());
+        } else if (authData.hasDataFromHttp()) {
+            // Authentication HTTP request. The format here should be compliant to RFC-6750
+            // (https://tools.ietf.org/html/rfc6750#section-2.1). Eg: Authorization: Bearer xxxxxxxxxxxxx
+            String httpHeaderValue = authData.getHttpHeader(HTTP_HEADER_NAME);
+            if (httpHeaderValue == null || !httpHeaderValue.startsWith(HTTP_HEADER_VALUE_PREFIX)) {
+                throw new AuthenticationException("Invalid HTTP Authorization header");
+            }
+
+            // Remove prefix
+            String token = httpHeaderValue.substring(HTTP_HEADER_VALUE_PREFIX.length());
+            return validateToken(token);
+        } else {
+            throw new AuthenticationException("No token credentials passed");
+        }
+    }
+
+    private static String validateToken(final String token) throws AuthenticationException {
+        if (StringUtils.isNotBlank(token)) {
+            return token;
+        } else {
+            throw new AuthenticationException("Blank token found");
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private DecodedJWT authenticateToken(final String token) throws AuthenticationException {
+        DecodedJWT jwt = JWT.decode(token);
+
+        Jwk jwk = null;
+        Algorithm algorithm = null;
+        try {
+            jwk = provider.get(jwt.getKeyId());
+            if(!(jwk.getPublicKey() instanceof RSAPublicKey)){
+                throw new JwtException("key needs to be a RSA Publickey");
+            }
+
+            algorithm = Algorithm.RSA256((RSAPublicKey) jwk.getPublicKey(), null);
```

Comment:
Does it only support this kind of algorithm? Are you considering getting the algorithm type from the alg field?

Reply:
well, I think HS256 is also a valid alg type even though rs256 is the default 😄 

I'll implement the feature 👍

Reply:
May I suggest that the `io.jsonwebtoken` library be used for token validation.  Pulsar already has the dependency. This will take care of applying the correct algorithm based on the provided key.  

It is easy to use, if you implement `io.jsonwebtoken.SigningKeyResolver` as a wrapper over the `UrlJwkProvider`.  Then you can instantiate a `io.jsonwebtoken.JwtParser` that correctly validates the token.

```
this.jwtParser = Jwts.parserBuilder()
                .setSigningKeyResolver(signingKeyResolver)
                .requireIssuer(issuerUrl)
                .build();
...
Jws<Claims> jwt = this.jwtParser.parseClaimsJws(tokenString);
// perform app-specific validations, e.g audience
```